hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java

```diff
@@ -805,10 +775,16 @@ private void startJoin() {
                     return;
                 }
 
-                // Run all joining members' provided pre join operations now, but only
-                //  execute them locally and on existing members of this cluster (do
-                //  not broadcast to other members joining within this batch)
-                runProvidedPostJoinOpsWithoutBroadcast();
+                if (syncJoinStrategy != null) {
+                    // Run all joining members' provided pre join operations now, but only
+                    //  execute them locally and on existing members of this cluster (do
+                    //  not broadcast to other members joining within this batch)
+                    runProvidedPostJoinOpsWithoutBroadcastSyncStrategyOnly();
+                } else {
+                    if (preJoinOperation != null) {
+                        nodeEngine.getOperationService().run(preJoinOperation);
```

Comment:
@JamesHazelcast is this ok to use plain `run` or should it be executed without broadcast?

eg `runOnJoinOpWithoutBroadcast(preJoinOperation)` where

```
    private void runOnJoinOpWithoutBroadcast(OnJoinOp onJoinOp) {
        try {
            onJoinOp.beforeRun();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
        onJoinOp.runWithoutBroadcastTo(joiningMembers.keySet());
    }

Reply:
I think it's okay since this `#runWithoutBroadcastTo` call was only used during batch joining, to prevent all joining members in that batch playing ping-pong - since we've removed batched joining entirely, we want to broadcast as normal since the state of other joining members isn't known during joining anymore.

Reply:
Agreed